frontend/public/components/routes.tsx

```diff
@@ -307,14 +312,270 @@ const filters = [{
   ],
 }];
 
-export const RoutesPage: React.SFC<RoutesPageProps> = props =>
-  <ListPage
+
+const getServicePorts = (service: object) => {
+  if (!service) {
+    return;
+  }
+
+  const ports = _.get(service, 'spec.ports', []);
+  const portOptions = _.map(ports, portMapping => {
+    return {
+      name: portMapping.name,
+      // \u2192 is a Unicode right arrow.
+      label: `${portMapping.port} \u2192 ${portMapping.targetPort} (${portMapping.protocol})`
+    };
+  });
+
+  return portOptions;
+};
+
+export class CreateRoute extends React.Component<null, CreateRouteState> {
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      stringData: {
+        namespace: getActiveNamespace(),
+        name: '',
+        hostname: '',
+        path: '',
+        service: '',
+        targetPort: '',
+      },
+      loaded: false,
+      error: '',
+      services: {},
+      selectedServicePorts: {},
+      // secure: false,
+      inProgress: false,
+    };
+  }
+
+  componentDidMount() {
+    k8sGet(ServiceModel, null, this.state.stringData.namespace)
+      .then(services => this.setState({
+        services: services.items,
+        loaded: true,
+      }))
+      .catch((err) => {
+        if (err.response.status !== 403) {
+          const errorMessage = err.message || 'Could not load services.';
+          this.setState({error: errorMessage});
+        }
+        return {
+          services: {}
+        };
+      });
+  }
+
+  handleChange: React.ReactEventHandler<HTMLInputElement> = event => {
+    this.setState({
+      stringData: {
```

Comment:
We chose `stringData` for secrets but it's a property on the Secret object. I'd probably call this `formData` or something like that